File: src/csharp/Intel/Iced.UnitTests/Intel/InstructionInfoTests/MemorySizeExtensionsTests.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if INSTR_INFO
using System;
using System.Collections.Generic;
using Iced.Intel;
using Xunit;

namespace Iced.UnitTests.Intel.InstructionInfoTests {
	public sealed class MemorySizeExtensionsTests {
		[Theory]
		[InlineData((MemorySize)(-1))]
		[InlineData((MemorySize)IcedConstants.MemorySizeEnumCount)]
		void GetInfo_throws_if_invalid_value(MemorySize memorySize) {
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.GetInfo());
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.GetSize());
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.GetElementSize());
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.GetElementType());
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.GetElementTypeInfo());
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.IsSigned());
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.IsPacked());
			Assert.Throws<ArgumentOutOfRangeException>(() => memorySize.GetElementCount());
			memorySize.IsBroadcast();// Does not throw
		}

		[Theory]
		[MemberData(nameof(VerifyMemorySizeProperties_Data))]
		void VerifyMemorySizeProperties(MemorySize memorySize, int size, int elementSize, MemorySize elementType, int elementCount, MemorySizeFlags flags) {
			var info = memorySize.GetInfo();
			Assert.Equal(memorySize, info.MemorySize);
			Assert.Equal(size, info.Size);
			Assert.Equal(elementSize, info.ElementSize);
			Assert.Equal(elementType, info.ElementType);
			Assert.Equal((flags & MemorySizeFlags.Signed) != 0, info.IsSigned);
			Assert.Equal((flags & MemorySizeFlags.Broadcast) != 0, info.IsBroadcast);
			Assert.Equal((flags & MemorySizeFlags.Packed) != 0, info.IsPacked);
			Assert.Equal(elementCount, info.ElementCount);

			Assert.Equal(size, memorySize.GetSize());
			Assert.Equal(elementSize, memorySize.GetElementSize());
			Assert.Equal(elementType, memorySize.GetElementType());
			Assert.Equal(elementType, memorySize.GetElementTypeInfo().MemorySize);
			Assert.Equal((flags & MemorySizeFlags.Signed) != 0, memorySize.IsSigned());
			Assert.Equal((flags & MemorySizeFlags.Packed) != 0, memorySize.IsPacked());
			Assert.Equal((flags & MemorySizeFlags.Broadcast) != 0, memorySize.IsBroadcast());
			Assert.Equal(elementCount, memorySize.GetElementCount());
		}
		public static IEnumerable<object[]> VerifyMemorySizeProperties_Data {
			get {
				foreach (var tc in MemorySizeInfoTestReader.GetTestCases())
					yield return new object[] { tc.MemorySize, tc.Size, tc.ElementSize, tc.ElementType, tc.ElementCount, tc.Flags };
			}
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/info/MemorySizeExtensionsTests.java
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.info;

import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

import com.github.icedland.iced.x86.MemorySize;
import com.github.icedland.iced.x86.MemorySizeInfo;
import com.github.icedland.iced.x86.internal.IcedConstants;

final class MemorySizeExtensionsTests {
	@ParameterizedTest
	@MethodSource("getInfo_throws_if_invalid_value_Data")
	void getInfo_throws_if_invalid_value(int memorySize) {
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.getInfo(memorySize));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.getSize(memorySize));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.getElementSize(memorySize));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.getElementType(memorySize));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.getElementTypeInfo(memorySize));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.isSigned(memorySize));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.isPacked(memorySize));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> MemorySize.getElementCount(memorySize));
		MemorySize.isBroadcast(memorySize);// Does not throw
	}

	static Iterable<Arguments> getInfo_throws_if_invalid_value_Data() {
		ArrayList<Arguments> result = new ArrayList<Arguments>();
		result.add(Arguments.of(-1));
		result.add(Arguments.of(IcedConstants.MEMORY_SIZE_ENUM_COUNT));
		return result;
	}

	@ParameterizedTest
	@MethodSource("verifyMemorySizeProperties_Data")
	void verifyMemorySizeProperties(int memorySize, int size, int elementSize, int elementType, int elementCount, int flags) {
		MemorySizeInfo info = MemorySize.getInfo(memorySize);
		assertEquals(memorySize, info.getMemorySize());
		assertEquals(size, info.getSize());
		assertEquals(elementSize, info.getElementSize());
		assertEquals(elementType, info.getElementType());
		assertEquals((flags & MemorySizeFlags.SIGNED) != 0, info.isSigned());
		assertEquals((flags & MemorySizeFlags.BROADCAST) != 0, info.isBroadcast());
		assertEquals((flags & MemorySizeFlags.PACKED) != 0, info.isPacked());
		assertEquals(elementCount, info.getElementCount());

		assertEquals(size, MemorySize.getSize(memorySize));
		assertEquals(elementSize, MemorySize.getElementSize(memorySize));
		assertEquals(elementType, MemorySize.getElementType(memorySize));
		assertEquals(elementType, MemorySize.getElementTypeInfo(memorySize).getMemorySize());
		assertEquals((flags & MemorySizeFlags.SIGNED) != 0, MemorySize.isSigned(memorySize));
		assertEquals((flags & MemorySizeFlags.PACKED) != 0, MemorySize.isPacked(memorySize));
		assertEquals((flags & MemorySizeFlags.BROADCAST) != 0, MemorySize.isBroadcast(memorySize));
		assertEquals(elementCount, MemorySize.getElementCount(memorySize));
	}

	public static Iterable<Arguments> verifyMemorySizeProperties_Data() {
		ArrayList<Arguments> result = new ArrayList<Arguments>();
		for (MemorySizeInfoTestCase tc : MemorySizeInfoTestReader.getTestCases())
			result.add(Arguments.of(tc.memorySize, tc.size, tc.elementSize, tc.elementType, tc.elementCount, tc.flags));
		return result;
	}
}
